File: app/src/main/java/droiddevelopers254/droidconke/viewmodels/AgendaViewModel.java
package droiddevelopers254.droidconke.viewmodels;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.ViewModel;

import droiddevelopers254.droidconke.datastates.AgendaState;
import droiddevelopers254.droidconke.repository.AgendaRepo;

public class AgendaViewModel extends ViewModel {
    private MediatorLiveData<AgendaState> agendaStateMediatorLiveData;
    private AgendaRepo agendaRepo;

    public AgendaViewModel(){
        agendaStateMediatorLiveData= new MediatorLiveData<>();
        agendaRepo = new AgendaRepo();
    }

    public LiveData<AgendaState> getAgendas(){
        return agendaStateMediatorLiveData;
    }
    public void fetchAgendas(){
        final LiveData<AgendaState> agendaStateLiveData= agendaRepo.getAgendaData();
        agendaStateMediatorLiveData.addSource(agendaStateLiveData,
                agendaStateMediatorLiveData ->{
                    if (this.agendaStateMediatorLiveData.hasActiveObservers()){
                        this.agendaStateMediatorLiveData.removeSource(agendaStateLiveData);
                    }
                    this.agendaStateMediatorLiveData.setValue(agendaStateMediatorLiveData);
                });
    }
}

File: app/src/main/java/droiddevelopers254/droidconke/viewmodels/AgendaViewModel.kt
package droiddevelopers254.droidconke.viewmodels

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import droiddevelopers254.droidconke.datastates.Result
import droiddevelopers254.droidconke.models.AgendaModel
import droiddevelopers254.droidconke.repository.AgendaRepo
import kotlinx.coroutines.launch

class AgendaViewModel(private val agendaRepo: AgendaRepo) : ViewModel() {
  private val agendaStateMediatorLiveData = MutableLiveData<List<AgendaModel>>()
  private val agendaError = MutableLiveData<String>()

  fun getAgendasResponse(): LiveData<List<AgendaModel>> = agendaStateMediatorLiveData

  fun getAgendaError(): LiveData<String> = agendaError

  fun fetchAgendas() {
    viewModelScope.launch {
      when (val value = agendaRepo.agendaData()) {
          is Result.Success -> agendaStateMediatorLiveData.postValue(value.data)
          is Result.Error -> agendaError.postValue(value.exception)
      }
    }
  }
}
